sublime_jedi/completion.py

```diff
@@ -146,61 +154,69 @@ def on_query_completions(self, prefix, locations):
         logger.info('JEDI completion triggered.')
 
         settings = get_settings(self.view)
-        only_jedi_completion = (
-            settings['sublime_completions_visibility'] in ('default', 'jedi')
-        )
-
-        previous_char = self.view.substr(locations[0] - 1)
         if settings['only_complete_after_regex']:
+            previous_char = self.view.substr(locations[0] - 1)
             if not re.match(settings['only_complete_after_regex'], previous_char):
                 return False
 
-        cplns = ask_daemon_with_timeout(
-            self.view,
-            'autocomplete',
-            location=locations[0],
-            timeout=settings['completion_timeout']
-        )
-
-        logger.info("Completion completed.")
-
-        cplns = [tuple(x) for x in self._sort_completions(cplns)]
-        logger.debug("Completions: {0}".format(cplns))
-
-        # disabled due to can't reproduce
-        # self._fix_tab_completion_issue()
-
-        if only_jedi_completion:
-            return cplns, PLUGIN_ONLY_COMPLETION
-        return cplns
+        with self._lock:
+            if self._last_location != locations[0]:
+                self._last_location = locations[0]
+                ask_daemon(
+                    self.view,
+                    self._receive_completions,
+                    'autocomplete',
+                    location=locations[0],
+                )
+                return [], PLUGIN_ONLY_COMPLETION
+
+            if self._last_location == locations[0] and self._completions:
+                self._last_location = None
+                return [
+                    tuple(x)
+                    for x in self._sort_completions(self._completions)
+                ]
+
+    def _receive_completions(self, view, completions):
+        if not completions:
+            return
+
+        logger.debug("Completions: {0}".format(completions))
+
+        with self._lock:
+            self._previous_completions = self._completions
+            self._completions = completions
+
+        if (completions and
+                not view.is_auto_complete_visible() or
+                not self._is_completions_subset()):
+            only_jedi_completion = (
+                get_settings(self.view)['sublime_completions_visibility']
+                in ('default', 'jedi')
+            )
+            view.run_command('hide_auto_complete')
+            view.run_command('auto_complete', {
+                'api_completions_only': only_jedi_completion,
+                'disable_auto_insert': True,
+                'next_completion_if_showing': False,
+            })
 
     def _sort_completions(self, completions):
         """Sort completions by frequency in document."""
         buffer = self.view.substr(sublime.Region(0, self.view.size()))
-
         return sorted(
             completions,
             key=lambda x: (
-                -buffer.count(x[1]),  # frequency in the text
+                -buffer.count(x[1]),
```

Comment:
please, keep comment as it is.
event now it's hard to say what is going on here

Reply:
should we move it in a function and five a better name ?

Reply:
I will add the comment again, about moving it out... I don't know